ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/CheckpointHeaderFetcher.java

```diff
@@ -66,6 +67,12 @@ public CheckpointHeaderFetcher(
       if (additionalHeaderCount == 0) {
         return completedFuture(singletonList(targetHeader));
       }
+    } else if (lastHeader.getNumber() + skip >= peer.chainState().getEstimatedHeight()) {
```

Comment:
To mimic our current approach.  What about making `CheckpointRange.end` an `Optional`?  In the case where `end` is empty, we can just download the next contiguous set of blocks from the previous header.

Reply:
Making end optional is going to take a fair bit of refactoring so I've started with just shifting to adjusting the skip value.  If the peer has re-orged so our previous checkpoint header is no longer on the canonical chain we won't be able to get any extra checkpoints from it or any extra headers if end was optional anyway.  Ultimately we'll wind up switching sync targets because we don't get new headers after a number of retries and things will continue ok.

Will take a look at making the end optional though and see what it winds up looking like.

Reply:
Making the end optional came out pretty well - it's not as isolated, spreading across 3 classes but I think that's reasonable and it's probably good to take the opportunity to potentially get a few more blocks past the sync target's chain head.